scrapinghub/utils.py

```diff
@@ -83,3 +87,81 @@ def proxy_methods(origin, successor, methods):
             successor_name, origin_name = method, method
         if not hasattr(successor, successor_name):
             setattr(successor, successor_name, getattr(origin, origin_name))
+
+
+def format_iter_filters(params):
+    """Format iter() filter param on-the-fly.
+
+    Support passing multiple filters at once as a list with tuples.
+    """
+    filters = params.get('filter')
+    if filters and isinstance(filters, list):
+        filter_data = []
+        for elem in params.pop('filter'):
+            if isinstance(elem, string_types):
+                filter_data.append(elem)
+            elif isinstance(elem, (list, tuple)):
+                filter_data.append(json.dumps(elem))
+            else:
+                raise ValueError(
+                    "Filter condition must be string, tuple or list")
+        if filter_data:
+            params['filter'] = filter_data
+    return params
+
+
+def parse_auth(auth):
+    """Parse authentification token.
+
+    >>> parse_auth(None)  # noqa
+    RuntimeError: No API key provided and SH_APIKEY environment variable not set
+    >>> os.environ['SH_APIKEY'] = 'apikey'
+    >>> parse_auth(None)
+    ('apikey', '')
+    >>> parse_auth(('user', 'pass'))
+    ('user', 'pass')
+    >>> parse_auth('user:pass')
+    ('user', 'pass')
+    >>> parse_auth('c3a3c298c2b8c3a6c291c284c3a9')
+    ('c3a3c298c2b8c3a6c291c284c3a9', '')
+    >>> parse_auth('312f322f333a736f6d652e6a77742e746f6b656e')
+    ('1/2/3', 'some.jwt.token')
+    """
+    if auth is None:
+        apikey = os.environ.get('SH_APIKEY')
+        if apikey is None:
+            raise RuntimeError("No API key provided and SH_APIKEY "
+                               "environment variable not set")
+        return (apikey, '')
+
+    if isinstance(auth, tuple):
+        if len(auth) != 2:
+            raise ValueError("Wrong amount of authentication credentials")
+        if not isinstance(auth[0], string_types):
+            raise ValueError("Login must me of a string type")
+        if not (auth[1] is None or isinstance(auth[1], string_types)):
+            raise ValueError("Password must be None or of a string type")
+        return auth
+
+    if not isinstance(auth, string_types):
+        raise ValueError("Wrong authentication credentials")
+
+    decoded_auth = None
+    try:
+        decoded_auth = decode(auth, 'hex_codec')
+    except (binascii.Error, TypeError):
+        login, _, password = auth.partition(':')
+        if not password:
+            raise ValueError("Bad apikey, please check your credentials")
+        return (login, password)
```

Comment:
> both apikey and jwt in my local tests pass this block without exception

That's right, both apikey and jwt are hex-encoded strings, so it should be decoded with `codecs.decode(auth, 'hex_codec')` on above. If there's `binascii.Error` or `TypeError`, we consider it as `user:password` pair, otherwise if there's no password - we assume that apikey wasn't correct.

Ok, let's try to split it to simplify the logic.